### PR TITLE
[QI2-1219] Create reusable workflows

### PR DESCRIPTION
**Pull request by oschusler**

### CODE REVIEW

File: .github/actions/poetry-setup/action.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+name: Install poetry
+description: Install python and poetry
```
Comment (Mythir) [inline .github/actions/poetry-setup/action.yaml:2]: And tox

File: .github/workflows/poetry-tests.yaml [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  unit-test:
+    uses: qutech-delft/github-actions/.github/workflows/_tox-unit-test.yaml@master
+    needs: [lint]
+    with:
+      python-version: ${{ inputs.python-version }}
+      poetry-version: ${{ inputs.poetry-version }}
+      os-versions: ${{ inputs.os-versions }}
+  e2e-test:
+    uses: qutech-delft/github-actions/.github/workflows/_custom-e2e-test.yaml@master
+    needs: [lint]
```
Comment (Mythir) [inline .github/workflows/poetry-tests.yaml:41]: IMO we should be running E2E test and unit tests parallel with lint. I presume you made them sequential to allow lint to find issues through static analysis that could also break the unit tests, but in practice 95% of the issues the linting tools find are formatting/typing stuff that don't impact unit tests. So I would say the speed up of running them all in parallel is worth it.

**Review by Mythir — CHANGES_REQUESTED** [review]
Looked through the actions/workflows, and also the subtasks still open on the board, and based on that I couldn't find any glaring issues. That said, I know GitHub actions has a lot of weird idiosyncrasies that make reviewing a bit tricky so the proof of the pudding is going to have to be in the eating I suppose.

File: .github/workflows/poetry-tests.yaml (hunk 2 above)
Comment (oschusler) [reply .github/workflows/poetry-tests.yaml:41]: The reason that we have lint as a first stage for all of our repositories, has to do with the duration of the unit tests. The linting is usually quite quick, and catches some of the obvious mistakes. The unit tests run on a number of environments, so take a long time (or cost more resources). Having the linting first is from the perspective of "fail fast"

File: .github/actions/poetry-setup/action.yaml (hunk 1 above)
Comment (oschusler) [reply .github/actions/poetry-setup/action.yaml:2]: Done

**Comment by oschusler** [comment]
> That said, I know GitHub actions has a lot of weird idiosyncrasies that make reviewing a bit tricky so the proof of the pudding is going to have to be in the eating I suppose.

I completely agree with you

File: .github/workflows/poetry-tests.yaml (hunk 2 above)
Comment (Mythir) [reply .github/workflows/poetry-tests.yaml:39]: I'm aware of the trade off with resources, but I wouldn't consider (what we call) the linting step to be quick in general. In platform it takes just a bit more time than pleasant. As for "fail fast", if you execute the linting step in parallel with the tests you can detect a failure just as quickly. And then you get even more information as you can also find faults in the unit tests and E2E tests. Either way, this not a huge deal, so I am willing to approve regardless, just consider it. Feel free to resolve this conversation by yourself if you don't want to run them in parallel.

**Review by Mythir — APPROVED** [review]

File: .github/workflows/poetry-tests.yaml (hunk 2 above)
Comment (oschusler) [reply .github/workflows/poetry-tests.yaml]: Updated to run in parallel. Can you approve again?

**Review by Mythir — APPROVED** [review]

**Pull request merged**